### ISSUE TITLE
Bug: SGEMM calls leak memory repeated in V1.4.1 and CLBlastClearCache() --Resource cleanup

### ISSUE THREAD

**Issue by TaihuLight** [CNugteren/CLBlast#336]
When I call CLBlastSgemm() and CLBlastHgemm(), the used memory increased rapidly with the number of calling these routine functions. Finally, it get -6 error code for OpenCL.
1. Test device and drivers
(1) HD graphic in Intel-6100U, Memory size = 12GB(DDR4), OS =Ubuntu 16.04.1 (kernel 4.15.0-39), OpenCL driver = SRB 5.0
(2) HD graphic in Intel 8700K, Memory size = 32GB(DDR4), OS =Ubuntu 16.04.5 (kernel 4.15.0-39), OpenCL driver = neo-18.45.11804 (https://github.com/intel/compute-runtime)
(3) Version of CLBLAST: released 1.4.1 and lasted master
2. Bug
Memory leak in calling GEMM routines. I try to use CLBlastClearCache() to clean resource, which can solve the memory leak but the speed of app is too low.

3. Test code
[clmleak.c.zip](https://github.com/CNugteren/CLBlast/files/2602264/clmleak.c.zip)

**Comment by CNugteren** [comment]
I'm afraid I can't reproduce your issue. I've taken the CLBlast C SGEMM sample and modified it as follows:
```
diff --git a/samples/sgemm.c b/samples/sgemm.c
index 92f3057..52360ee 100644
--- a/samples/sgemm.c
+++ b/samples/sgemm.c
@@ -62,6 +62,8 @@ int main(void) {
   cl_command_queue queue = clCreateCommandQueue(context, device, 0, NULL);
   cl_event event = NULL;
 
+  while(1) {
+
   // Populate host matrices with some example data
   float* host_a = (float*)malloc(sizeof(float)*m*k);
   float* host_b = (float*)malloc(sizeof(float)*n*k);
@@ -99,14 +101,15 @@ int main(void) {
   printf("Completed SGEMM with status %d\n", status);
 
   // Clean-up
-  free(platforms);
-  free(devices);
   free(host_a);
   free(host_b);
   free(host_c);
   clReleaseMemObject(device_a);
   clReleaseMemObject(device_b);
   clReleaseMemObject(device_c);
+  }
+  free(platforms);
+  free(devices);
   clReleaseCommandQueue(queue);
   clReleaseContext(context);
   return 0;
```

I ran it on these two Intel CPUs under Debian Linux for a while and saw that memory remains constant:
```
Platform #1: Experimental OpenCL 2.1 CPU Only Platform
 `-- Device #0: Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
Platform #3: Intel(R) OpenCL
 `-- Device #0: Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
```

I'm not sure how to measure memory usage on the GPU, other than regular top since I guess DRAM is shared between CPU and GPU on this platform? Anyway, ran it with Intel Neo and Beignet for a bit and observed no crashes.

**Comment by TaihuLight** [comment]
@CNugteren 
(1) As DRAM is shared for intel CPU and its GPU, the memory usage can be measured with top - p PID(specific ID for progress call SGEMM).
(2) I will test samples/sgemm.c from your repo, it also get CLBLast: OpenCL error: clbuildProgram: -6  after run 999368 times on the GPU in intel 8700K with neo.

**Comment by CNugteren** [comment]
I have taken your example code and run it on my machine, and indeed the memory usage increases. However, when I move the call to your `initIntelGPU` outside of the loop (and the 3 related clean-up lines), the memory remains constant. I have increased the loop counter to 1M, no issues. This is with Intel Neo on the GPU by the way.

So I don't think the 'issue' is CLBlast related. And anyway in a real application you wouldn't create that many OpenCL devices/platforms/queues/contexts, right?

**Comment by CNugteren** [comment]
OK, so the 'issue' with the memory leaks is solved then.

Now you are having an issue with speed. After you've run the tuners for those sizes, did you use those new found parameters? Note that the tuning script on itself doesn't change the CLBlast code or something, see the docs on tuning. What do the two tuners report as final best numbers? How does that compare to their counter-parts with `-precision 32`?

For example, with my Intel Neo OpenCL on a GPU, I did a small test with the first part of tuning (didn't bother with the rest to safe time), and I found:
```
./clblast_tuner_xgemm_direct -precision 16 -m 384 -n 384 -k 1024 -alpha 1.0 -beta 0.0
(...)
* Found 41 configuration(s)
(...)
* Got average result of 11.24 ms: 26.9 GFLOPS
* Found best result 4.61 ms: 65.5 GFLOPS
* Best parameters: KWID=2 MDIMAD=16 MDIMCD=16 NDIMBD=8 NDIMCD=8 PADA=1 PADB=1 PRECISION=16 VWMD=2 VWND=4 WGD=32
* Writing a total of 41 results to 'clblast_xgemm_direct_1_16.json'
(...)
```
and:
```
./clblast_tuner_xgemm_direct -precision 32 -m 384 -n 384 -k 1024 -alpha 1.0 -beta 0.0
(...)
* Found 41 configuration(s)
(...)
* Got average result of 34.72 ms: 8.7 GFLOPS
* Found best result 4.96 ms: 60.9 GFLOPS
* Best parameters: KWID=2 MDIMAD=16 MDIMCD=16 NDIMBD=8 NDIMCD=8 PADA=1 PADB=1 PRECISION=32 VWMD=1 VWND=4 WGD=32
* Writing a total of 41 results to 'clblast_xgemm_direct_1_16.json'
(...)
```
So it seems in that simplified case with your specific parameters there is not much to gain from half-precision (60.9 vs 65.5 GFLOPS) with the kernel implementation of the direct GEMM kernel in CLBlast.

**Comment by TaihuLight** [comment]
@CNugteren  First, thank you for your timely reply.
(1) I can get the similar results after run the tuners, and I get half-precision for my specific matrices can only get very small speedups on HD 630 Graphics with the NEO driver. But it should not correct that HGEMM routines are slower than SGEMM routines for matrices of the same input sizes, which are only  different in the data type of elements. 
I want to know  why this problem is happened? It is wrong with Intel GPU driver software or CLBlast lib? 
(2) I also test HGEMM and SGEMM for matrices (e.g., [342, 1089, 364]) of the the same input sizes on HD Graphics Skylake ULT GT2 in Intel i-6700 Processor with SRB 5.0 and Ubuntu 16.04.5 (patched kernel 4.7), it can get better speedups (e.g., 1.2x-1.7x), but the executing time of HGEMM routines for some specific matrices increases by nearly 2-3 times compared with SGEMM routines on matrices of the same input sizes. These results are very strange, I want to know why?

**Comment by CNugteren** [comment]
Thanks for the detailed reply. A couple of remarks from my side (numbers don't match your comments):

1. When you say `it should not correct that HGEMM routines are slower than SGEMM routines for matrices of the same input sizes, which are only different in the data type of elements.`, you are not correct. Because FP16 is only 2 bytes and not 4, this changes memory behaviour significantly. Some devices work best if each thread accesses subsequent 4-byte segments of memory (e.g. NVIDIA GPUs), requiring at least 2-wide vector loads for FP16. This is just an example, but memory behaviour can change and make certain configurations less favourable for FP16 compared to FP32.

2. If you are in doubt that your benchmarks are correct, I suggest just using the included `clblast_client_xgemm` benchmark tool. That way we both know that we are testing the same thing.

3. From the tuner results it seems like you get some benefit of FP16 over FP32, but not too much. However, your benchmarks show you get worse performance. Now, after you've verified this (see point 2 above), you have to realize CLBlast's GEMM is built as either running a couple of kernels in a row (with the GEMM indirect kernel being the main one) or a single kernel (the GEMM direct kernel). Thus, there might be other kernels involved which are slow. If you compile CLBlast with `-DVERBOSE=ON` (CMake argument), you'll get a lot more debug timing output to figure out what is going on.

4. If you are not happy with the choice of kernel it make (e.g. direct or indirect), you can also play with the [the `XGEMM_MIN_INDIRECT_SIZE` parameter](https://github.com/CNugteren/CLBlast/blob/master/src/database/kernels/gemm_routine/gemm_routine_16.hpp), or just run the GEMM routine tuner after you've tuner the kernels themselves.

**Comment by TaihuLight** [comment]
When I compile CLBlast with -DVERBOSE=ON, I get debug timing output as following:
(1) Timing with clock(), which is similar to std::chrono::duration that is used in CLBlast  (https://github.com/CNugteren/CLBlast/blob/master/src/routines/common.cpp)
```
  // Prints the elapsed execution time in case of debugging in verbose mode
  #ifdef VERBOSE
    queue.Finish();
    const auto elapsed_time = std::chrono::steady_clock::now() - start_time;
    printf("[DEBUG] Completed kernel in %.2lf ms\n", timing);
#endif
```
```
[DEBUG] GEMM_Copy_8_8_2_1_Pad_32_8_4_2_Transpose_16_0_0_4_Padtranspose_1_16_2_Xgemm_0_1_32_2_16_16_64_8_8_64_1_1_0_0_4_4_XgemmDirect_2_8_8_8_8_1_1_1_1_8_GemmRoutine_128
[DEBUG] Running kernel 'CopyPadMatrix'
[DEBUG] Completed kernel in 0.54 ms
[DEBUG] Running kernel 'CopyPadMatrix'
[DEBUG] Completed kernel in 0.33 ms
[DEBUG] Running kernel 'Xgemm'
[DEBUG] Completed kernel in 2.65 ms
[DEBUG] Running kernel 'TransposeMatrix'
[DEBUG] Completed kernel in 0.20 ms
No of runing is 2, kernel executing time is 3.183000 ms
```
**My SGEMM operations involve four kernels as above, but timing is not accurate, e.g., 3.183 < 3.73 = 0.54 + 0.33 + 2.65 + 0.20**  
(2) Timing with time return from event related to SGEMM().
```
double get_event_exec_time(cl_event event) {
	cl_ulong start_time, end_time;
	/*Get start device counter for the event*/
	clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_START, sizeof(cl_ulong),
			&start_time, NULL);
	/*Get end device counter for the event*/
	clGetEventProfilingInfo(event, CL_PROFILING_COMMAND_END, sizeof(cl_ulong),
			&end_time, NULL);
	/*Convert the counter values to milli seconds*/
	double total_time = (end_time - start_time) * 1e-6;
	return total_time;
}

// call  get_event_exec_time(cl_event event)  to timing for SGEMM
        clFinish(queue); 
	cl_event sevent = NULL;
	CLBlastStatusCode status = CLBlastSgemm(CLBlastLayoutRowMajor,
			CLBlastTransposeYes, CLBlastTransposeNo, m, n, k, alpha, buf_a, 0,
			a_ld, buf_b, 0, b_ld, beta, buf_c, 0, c_ld, &queue, &sevent);

	clFinish(queue);  //wait for finishing the command_queue
	double kerneltime = get_event_exec_time(sevent);
```
It gets the following results, and kernel execution time is too small. Although this timing method can get better speedup for SGEMM over HGEMM, it is correct for timing kernel executing timing?
```
[DEBUG] GEMM_Copy_8_8_2_1_Pad_32_8_4_2_Transpose_16_0_0_4_Padtranspose_1_16_2_Xgemm_0_1_32_2_16_16_64_8_8_64_1_1_0_0_4_4_XgemmDirect_2_8_8_8_8_1_1_1_1_8_GemmRoutine_128
[DEBUG] Running kernel 'CopyPadMatrix'
[DEBUG] Completed kernel in 0.55 ms
[DEBUG] Running kernel 'CopyPadMatrix'
[DEBUG] Completed kernel in 0.30 ms
[DEBUG] Running kernel 'Xgemm'
[DEBUG] Completed kernel in 2.54 ms
[DEBUG] Running kernel 'TransposeMatrix'
[DEBUG] Completed kernel in 0.19 ms
No of runing is 2, kernel executing time is 0.057333 ms
```

**Comment by CNugteren** [comment]
I'll take a look later at what you are doing in details, but I'm not sure it is relevant:

* If you want to report speed issues of CLBlast, then use the CLBlast clients to time your code (e.g. run `./clblast_client_xgemm --precision 16` and compare it to `./clblast_client_xgemm --precision 32` with whatever matrix sizes you want).

* If you want a deeper understanding to figure out why the kernel tuner reports favorable performance of HGEMM vs SGEMM kernels, but the end result is the other way around, you can do some debugging with `-DVERBOSE=ON`. This is meant to show that other kernels are running and show how fast they go. Actual total timing might be different, because the verbose timing measurement adds code, including event synchronisation.

**Comment by TaihuLight** [comment]
Compared with released CLBlast v1.4.1, the version of master (bca1506) has better performance for HGEMM on Intel HD 630 Graphics in 8700k processors. The tuning results are shown in following:
```
./clblast_tuner_xgemm_direct -precision 16 -m 384 -n 384 -k 1024 -alpha 1.0 -beta 0.0
* Got average result of 31.32 ms: 9.6 GFLOPS
* Found best result 2.68 ms: 112.8 GFLOPS
* Best parameters: KWID=16 MDIMAD=8 MDIMCD=8 NDIMBD=16 NDIMCD=16 PADA=1 PADB=1 PRECISION=16 VWMD=1 VWND=1 WGD=32
```

**Comment by CNugteren** [comment]
That seems similar to the results you posted earlier, right? Anyway, not sure how this is related to the issue you are having with not seeing the performance you expect when using CLBlast compared to when running the tuners.

**Comment by TaihuLight** [comment]
I can get better performance (e.g., HGEMM gets 1.3 - 1.7 times speedup over SGEMM) using CLBlast  master (bca1506) instead of released version 1.4.1.  Why this is happen since I test it on the same GPU device (Intel HD 630 Graphics in 8700k processors) and the same Intel OpenCL GPU drivers (neo-18.45.11804)? And this result is correct ?

**Comment by CNugteren** [comment]
Master branch is also well rested, soon to be released as 1.5.0. you can check the [changelog](https://github.com/CNugteren/CLBlast/blob/master/CHANGELOG).

Anyway, is there still an issue or not? Otherwise I'll close this.

**Comment by TaihuLight** [comment]
Thank you, please close this issue.
